pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1418,3 +1436,258 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
+}
+
+// collapse all zones in same region. Join FailureDomain with well known separator
+func (vs *VSphere) collapseZonesInRegion(ctx context.Context, zones []cloudprovider.Zone) ([]cloudprovider.Zone, error) {
+	// first create a map of region -> list of zones in that region
+	regionToZones := make(map[string][]string)
+	for _, zone := range zones {
+		fds, exists := regionToZones[zone.Region]
+		if !exists {
+			fds = make([]string, 0)
+		}
+		regionToZones[zone.Region] = append(fds, zone.FailureDomain)
+	}
+
+	// Join all fds in same region and return Zone instances
+	collapsedZones := make([]cloudprovider.Zone, 0)
+	for region, fds := range regionToZones {
+		fdSet := sets.NewString(fds...)
+		appendedZone := volumeutil.ZonesSetToLabelValue(fdSet)
+		collapsedZones = append(collapsedZones, cloudprovider.Zone{FailureDomain: appendedZone, Region: region})
+	}
+	return collapsedZones, nil
+}
+
+// GetZonesForDatastore returns all the zones from which this datastore is visible
+func (vs *VSphere) GetZonesForDatastore(ctx context.Context, datastore string) ([]cloudprovider.Zone, error) {
+	vsi, err := vs.getVSphereInstanceForServer(vs.cfg.Workspace.VCenterIP, ctx)
+	if err != nil {
+		klog.Errorf("Failed to get vSphere instance: %+v", err)
+		return nil, err
+	}
+	dc, err := vclib.GetDatacenter(ctx, vsi.conn, vs.cfg.Workspace.Datacenter)
+	if err != nil {
+		klog.Errorf("Failed to get datacenter: %+v", err)
+		return nil, err
+	}
+	// get the hosts mounted on this datastore
+	// datastore -> ["host-1", "host-2", "host-3", ...]
+	ds, err := dc.GetDatastoreByName(ctx, datastore)
+	if err != nil {
+		klog.Errorf("Failed to get datastore by name: %v: %+v", datastore, err)
+		return nil, err
+	}
+	dsHosts, err := ds.GetDatastoreHostMounts(ctx)
+	if err != nil {
+		klog.Errorf("Failed to get datastore host mounts for %v: %+v", datastore, err)
+		return nil, err
+	}
+	klog.V(4).Infof("Got host mounts for datastore: %v: %v", datastore, dsHosts)
+
+	// compute map of zone to list of hosts in that zone across all hosts in vsphere
+	// zone -> ["host-i", "host-j", "host-k", ...]
+	zoneToHosts, err := vs.GetZoneToHosts(ctx, vsi)
+	if err != nil {
+		klog.Errorf("Failed to get zones for hosts: %+v", err)
+		return nil, err
+	}
+	klog.V(4).Infof("Got zone to hosts: %v", zoneToHosts)
+
+	// datastore belongs to a zone if all hosts in that zone mount that datastore
+	dsZones := make([]cloudprovider.Zone, 0)
+	for zone, zoneHosts := range zoneToHosts {
+		// if zone is valid and zoneHosts is a subset of dsHosts, then add zone
+		if zone.Region != "" && containsAll(dsHosts, zoneHosts) {
+			dsZones = append(dsZones, zone)
+		}
+	}
+	klog.V(4).Infof("Datastore %s belongs to zones: %v", datastore, dsZones)
+	return dsZones, nil
+}
+
+// GetZoneToHosts returns a map of 'zone' -> 'list of hosts in that zone' in given VC
+func (vs *VSphere) GetZoneToHosts(ctx context.Context, vsi *VSphereInstance) (map[cloudprovider.Zone][]vmwaretypes.ManagedObjectReference, error) {
```

Comment:
@subramanian-neelakantan 
There is a lot going on here.  I would feel better if this was unit tested.

Reply:
I have added unit tests to cover this now. Can you kindly take another look?